File: data_structures/trie.java
// A Trie is a tree-like data structure that stores a dynamic set of strings, where the keys are usually strings.
// It is used to store words in a way that facilitates fast prefix-based searches.
// Each node in a trie represents a single character, and each path down the tree represents a word.

// Imagine you are designing an autocomplete feature for a search engine.
// As a user types in the search box, you want to suggest possible completions based on the current prefix.
// A trie is an ideal data structure for this task because it allows you to quickly find all words that start with the given prefix.
// For example, if a user types "cat", the trie can quicky suggest "cat", "cater", "catapult" et cetera, based on the stored words.

// Use Cases
// 1. Autocomplete Systems: Suggesting words based on prefixes.
// 2. Spell checkers: Validating if a string is a valid word.
// 3. IP Routing: Longest prefix matching in networking.
// 4. Data compression: Strong dictionary words in a compact format.

// A trie is typically implemented with a root node that branches into child nodes, each representing a character.
// A node may have multiple children, corresponding to the next possible characters in the words.

import java.util.HashMap;
import java.util.Map;

class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    boolean isEndOfWord = false;

    public TrieNode() {}
}

File: data_structures/trie.kt
// A trie is a tree-like data structure that stores a dynamic set of strings.
// It's useful for implementing dictionaries with efficient search, insert and delete operations.

// Think of a contact list on your phone with thousands of names.
// Every time you start typing a name in the search bar, the phone suggests possible matches.
// This can be implemented using a trie.
// Each node of a trie represents a character, and a path from the root to a node represents a string or a prefix of a string.
// The end of a way is usually marked with a special end-of-word marker.

// With tries:
// Lookup operations are faster than with other data structures like hash tables.
// They can be used for auto-suggestions.
// They can efficiently retrieve all keys with a common prefix.

class TrieNode {
    val children = mutableMapOf<Char, TrieNode>()
    val isEndOfWord = false
}

class Trie {
    private val root = TrieNode()

    fun insert(word: String) {
        val currentNode = root
        for (ch in word) {
            currentNode = currentNode.children.computeIfAbsent(ch) { TrieNode() }
        }
        currentNode.isEndOfWord = true
    }

    fun search(word: String): Boolean {
        var currentNode = root
        for (ch in word) {
            currentNode = currentNode.children[ch] ?: return false
        }
        return currentNode.isEndOfWord
    }

    fun startsWith(prefix: String): Boolean {
        var currentNode = root
        for (ch in prefix) {
            currentNode = currentNode.children[ch] ?: return false
        }
        return true
    }
}

fun main() {
    val trie = Trie()
    trie.insert("apple")
    trie.insert("app")

    println(trie.search("apple")) // true
    println(trie.search("app")) // true
    println(trie.search("appl")) // false
    println(trie.startsWith("app")) // true
    println(trie.startsWith("apl")) // false
}